R/iterators.R

```diff
@@ -517,30 +517,33 @@ simple_vs_index <- function(x, i, na_ok = FALSE) {
 #' V(g)[.env$x]
 #'
 `[.igraph.vs` <- function(x, ..., na_ok = FALSE) {
-  args <- lazy_dots(..., .follow_symbols = FALSE)
+  args <- rlang::enquos(...)
 
-  ## If indexing has no argument at all, then we still get one,
-  ## but it is "empty", a name that is  ""
+  ## Special case, no argument
+  if (length(args) == 0) {
+    return(x)
+  }
 
-  ## Special case, no argument (but we might get an artificial
-  ## empty one
-  if (length(args) < 1 ||
-    (length(args) == 1 && inherits(args[[1]]$expr, "name") &&
-      as.character(args[[1]]$expr) == "")) {
+  ## or empty (what is the test case for this??)
+  ## try() because of .env$something would otherwise error
+  first_is_empty <- !nzchar(paste(rlang::quo_get_expr(args[[1]]), collapse = ""))
+  if (length(args) == 1 && first_is_empty) {
     return(x)
   }
 
   ## Special case: single numeric argument
-  if (length(args) == 1 && inherits(args[[1]]$expr, "numeric")) {
```

Comment:
This case is for literal numeric values and can be kept as is? What about integer values?

``` r
rlang::expr(5L) |> inherits("numeric")
#> [1] FALSE
```

<sup>Created on 2024-08-22 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

We can keep most of this logic as is also when using rlang:

``` r
library(rlang)

tester <- function(...) {
  quos <- enquos(...)

  exprs <- purrr::map(quos, quo_get_expr)
  purrr::map(exprs, class)
}

tester(5, symbol, x + 1)
#> [[1]]
#> [1] "numeric"
#> 
#> [[2]]
#> [1] "name"
#> 
#> [[3]]
#> [1] "call"
```

<sup>Created on 2024-08-22 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>